src/org/traccar/protocol/EelinkProtocolDecoder.java

```diff
@@ -209,41 +325,72 @@ protected Object decode(
 
         ChannelBuffer buf = (ChannelBuffer) msg;
 
-        buf.skipBytes(2); // header
-        int type = buf.readUnsignedByte();
-        buf.readShort(); // length
-        int index = buf.readUnsignedShort();
+        // check that incoming protocall has correct eelink prefix
+        if (buf.readShort() != HEADER_KEY) {
+            Log.warning(new IllegalArgumentException("Invalid Header"));
+            return null;
+        }
+
+        int type = buf.readUnsignedByte(); // protocol number
 
-        if (type != MSG_GPS && type != MSG_DATA) {
-            sendResponse(channel, type, index);
+        short length = buf.readShort(); // length
+        if (buf.readableBytes() != length) { // check remaining length
+            Log.warning(new IllegalArgumentException("Packet length error"));
+            return null;
         }
 
-        if (type == MSG_LOGIN) {
+        index = buf.readUnsignedShort();
+
+        // check if we have received an old packet
+        /*int newIndex = buf.readUnsignedShort(); // current message index
```

Comment:
But that may still end up doing something, ie say someone did an action based on a received status change, but if they saw it twice when the second was just a duplicate it could cause a problem.
Personally I cant see the harm if the option to have it in the decoder if index variable persists across each packet received.

ie this unit can forward SMS received to the server, you only want to process these once if that action was designed say to top up the credit.